### PR TITLE
Bug 852583 - enable bluetooth emulation

### PR DESCRIPTION
**Pull request by vicamo**

### CODE REVIEW

File: arch/arm/configs/goldfish_defconfig [hunk 1]
```diff
@@ -1235,9 +1349,11 @@ CONFIG_GENERIC_FIND_LAST_BIT=y
...
-CONFIG_ZLIB_INFLATE=y
-CONFIG_ZLIB_DEFLATE=y
+CONFIG_LIBCRC32C=y
+CONFIG_TEXTSEARCH=y
+CONFIG_TEXTSEARCH_KMP=y
+CONFIG_TEXTSEARCH_BM=y
+CONFIG_TEXTSEARCH_FSM=y
 CONFIG_PLIST=y
 CONFIG_HAS_IOMEM=y
 CONFIG_HAS_IOPORT=y
```
Comment (tdz) [inline arch/arm/configs/goldfish_defconfig:1359]: This change to the config files does a lot. Most of it looks unrelated to Bluetooth.

I guess most of these changes come from simply updating the default config with something that 'make config' produced. I suggest to move the 'make config'-related changes to a separate commit and put the Bluetooth changes in top of it.

File: arch/arm/mach-goldfish/rfkill.c [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+};
+
+struct goldfish_rfkill {
+	void __iomem *base;
+	u32 irq;
+	struct rfkill *rfk;
+	struct tasklet_struct irq_tasklet;
+	spinlock_t lock;
+
+	u8 block_mask;
```
Comment (tdz) [inline arch/arm/mach-goldfish/rfkill.c]: Nit: Add a comment to say it's protected  by 'lock'

File: arch/arm/mach-goldfish/rfkill.c [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+	spin_lock_irqsave(&drvdata->lock, irq_flags);
+
+	hb = readl(drvdata->base + OFFSET_HWBLOCK);
+	if (hb && !(drvdata->block_mask & STATE_HWBLOCKED)) {
+		drvdata->block_mask |= STATE_HWBLOCKED;
+	} else if (!hb && (drvdata->block_mask & STATE_HWBLOCKED)) {
+		drvdata->block_mask &= ~STATE_HWBLOCKED;
+	} else {
+		goto interrupt_done;
```
Comment (tdz) [inline arch/arm/mach-goldfish/rfkill.c]: This is a shared interupt handler, so it's supposed to check whether its device really triggered the interupt. If that's not the case it's supposed to return IRQ_NONE.

I suggest to declare 'int ret = IRQ_HANDLED' at the top of the function, assign 'ret = IRQ_NONE' before this goto statement, and return the value of ret below.

File: arch/arm/mach-goldfish/rfkill.c [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+	int ret;
+
+	drvdata = kzalloc(sizeof(*drvdata), GFP_KERNEL);
+	if (drvdata == NULL) {
+		return -ENOMEM;
+	}
+
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+	if (res == NULL) {
+		ret = -EINVAL;
```
Comment (tdz) [inline arch/arm/mach-goldfish/rfkill.c]: EINVAL is for function arguments. I suggest using ENXIO.

File: arch/arm/mach-goldfish/rfkill.c [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+		goto err_get_resource;
+	}
+#ifdef CONFIG_ARM
```
Comment (tdz) [inline arch/arm/mach-goldfish/rfkill.c]: Why is ARM special wrt. to IO remapping?

File: arch/arm/mach-goldfish/rfkill.c [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+	drvdata->base = (void __iomem *)IO_ADDRESS(res->start - IO_START);
+#else
+	drvdata->base = ioremap(res->start, 0x1000);
```
Comment (tdz) [inline arch/arm/mach-goldfish/rfkill.c]: I suggest using resource_size(res) for the length here.

File: arch/arm/mach-goldfish/rfkill.c [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+#endif
+
+	res = platform_get_resource(pdev, IORESOURCE_IRQ, 0);
+	if (res == NULL) {
+		ret = -EINVAL;
+		goto err_get_resource;
+	}
+	drvdata->irq = res->start;
+
```
Comment (tdz) [inline arch/arm/mach-goldfish/rfkill.c]: There is 'platform_get_irq', which does the check, or returns the correct error code otherwise.

File: arch/arm/mach-goldfish/rfkill.c [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+	rfkill_free(rfk);
+err_rfk_allocate:
+	free_irq(drvdata->irq, pdev);
+err_request_irq:
+err_get_resource:
+	kfree(drvdata);
+	return ret;
+}
+
+static int __devexit goldfish_rfkill_remove(struct platform_device *pdev)
```
Comment (tdz) [inline arch/arm/mach-goldfish/rfkill.c]: You should run iounmap() for drvdata->base in this function.

File: arch/arm/mach-goldfish/rfkill.c [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+{
+	struct goldfish_rfkill *drvdata = platform_get_drvdata(pdev);
+
+	// Disable interrupt forever.
+	writel(CMD_DISABLE_INT, drvdata->base + OFFSET_CMD);
```
Comment (tdz) [inline arch/arm/mach-goldfish/rfkill.c]: I think it can happen that you have a running tasklet while executing this line. This might re-enable the interrupt concurrently. I think the simplest solution is to disable the interupt a second time after tasklet_kill. Maybe this is handled implicitly by the writel() call in goldfish_rfkill_toggle_radio().

File: arch/arm/mach-goldfish/rfkill.c [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+
+	tasklet_kill(&drvdata->irq_tasklet);
+
+	goldfish_rfkill_toggle_radio(drvdata, RFKILL_STATE_OFF);
+	rfkill_unregister(drvdata->rfk);
+
+	free_irq(drvdata->irq, pdev);
+
+	platform_set_drvdata(pdev, NULL);
```
Comment (tdz) [inline arch/arm/mach-goldfish/rfkill.c]: Is this necessary?

File: arch/arm/mach-goldfish/rfkill.c [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+static void __exit goldfish_rfkill_exit(void)
+{
+	platform_driver_unregister(&goldfish_rfkill_driver);
+}
+
+module_init(goldfish_rfkill_init);
+module_exit(goldfish_rfkill_exit);
+
```
Comment (tdz) [inline arch/arm/mach-goldfish/rfkill.c]: MODULE_AUTHOR, MODULE_LICENSE and MODULE_DESCRIPTION seem missing.

File: arch/arm/configs/goldfish_defconfig (hunk 1 above)
Comment (vicamo) [inline arch/arm/configs/goldfish_defconfig:1359]: It is. That's the first commit.

File: arch/arm/mach-goldfish/rfkill.c (hunk 11 above)
Comment (vicamo) [inline arch/arm/mach-goldfish/rfkill.c]: @tdz Thanks for your review. Every time you teach me a lot. Really!

File: arch/arm/mach-goldfish/rfkill.c [hunk 12]
```diff
@@ -0,0 +1,253 @@
+/*
+ * Bluetooth built-in chip control
+ *
+ * Copyright (c) 2013 You-Sheng Yang
```
Comment (tdz) [inline arch/arm/mach-goldfish/rfkill.c]: Michael just made me aware that the copyright holder here should be 'Mozilla Foundation'.

File: arch/arm/mach-goldfish/rfkill.c (hunk 5 above)
Comment (vicamo) [inline arch/arm/mach-goldfish/rfkill.c]: That's old goldfish code. I'll follow new coding style used in goldfish drivers landed recently in kernel 3.9.

File: arch/arm/mach-goldfish/rfkill.c (hunk 8 above)
Comment (vicamo) [inline arch/arm/mach-goldfish/rfkill.c]: Like previous one, I'll use devm_\* functions to manage device resources automatically.

File: arch/arm/mach-goldfish/rfkill.c (hunk 12 above)
Comment (vicamo) [inline arch/arm/mach-goldfish/rfkill.c]: Sure.

File: arch/arm/mach-goldfish/rfkill.c (hunk 10 above)
Comment (vicamo) [inline arch/arm/mach-goldfish/rfkill.c]: Probably not. Removed.

**Comment by vicamo** [comment]
Address all review comments but https://github.com/mozilla-b2g/kernel_goldfish/pull/1#discussion_r3729771

**Comment by vicamo** [comment]
1) Removed x86 parts because it doesn't actually work.
2) Rewrite IRQ handling
3) Support multiple RFKILL devices but only initialize bluetooth one now.

@tdz , could you have a review? Thank you.

File: drivers/net/goldfish_rfkill.c [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+
+struct goldfish_rfkill_priv;
+struct rfkill_data {
+	struct goldfish_rfkill_priv *priv;
+	struct rfkill *rfkill;
+	enum rfkill_type type;
+};
+
+struct goldfish_rfkill_priv {
+	void __iomem *base;
```
Comment (tdz) [inline drivers/net/goldfish_rfkill.c:45]: I suggest making this of type u8 instead of void.

File: drivers/net/goldfish_rfkill.c [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+
+	struct rfkill_data data[RFKILL_TYPE_MAX];
+
+	u32 sw_block;
+	u32 hw_block;
+	u32 hw_mask;
+};
+
+#define goldfish_rfkill_read32(priv, ofs)      readl((priv)->base + (ofs))
+#define goldfish_rfkill_write32(priv, v, ofs)  writel((v), (priv)->base + (ofs))
```
Comment (tdz) [inline drivers/net/goldfish_rfkill.c]: I think readl and writel are deprecated in favor of ioread32 and iowrite32.

File: drivers/net/goldfish_rfkill.c [hunk 15]
```diff
@@ -0,0 +1,336 @@
...
+
+	/* Discover which interrupts are active/pending */
+	inta = goldfish_rfkill_read_irq(priv);
+
+	/* Ignore interrupt if there's nothing to service. */
+	if (!inta) {
+		/* Re-enable interrupt again if necessary. */
+		if (test_bit(STATUS_ENABLE_INTERRUPT, &priv->status))
+			goldfish_rfkill_enable_irq(priv);
+		goto isr_done;
```
Comment (tdz) [inline drivers/net/goldfish_rfkill.c]: It's probably just a matter of taste, but I would replace the goto by an if-else branch here.

File: drivers/net/goldfish_rfkill.c [hunk 16]
```diff
@@ -0,0 +1,336 @@
...
+	priv->base = devm_ioremap(&pdev->dev, r->start, resource_size(r));
+	if (priv->base == NULL) {
+		dev_err(&pdev->dev, "devm_ioremap failed\n");
+		return -ENOMEM;
+	}
+
+	priv->irq = platform_get_irq(pdev, 0);
+	if (priv->irq < 0) {
+		dev_err(&pdev->dev, "platform_get_irq failed\n");
+		return -EINVAL;
```
Comment (tdz) [inline drivers/net/goldfish_rfkill.c]: The value of priv->irq will contain an error code. You should return it.

File: drivers/net/goldfish_rfkill.c [hunk 17]
```diff
@@ -0,0 +1,336 @@
...
+	}
+
+	goldfish_rfkill_write_block(priv, priv->hw_block | priv->sw_block);
+
+toggle_radio_done:
+	spin_unlock_irqrestore(&priv->irq_lock, irq_flags);
+	return ret;
+}
+
+static int goldfish_rfkill_init_rfkill(struct goldfish_rfkill_priv *priv,
```
Comment (tdz) [inline drivers/net/goldfish_rfkill.c:188]: This function is only executed if CONFIG_BT is set, so it should be enclosed by CONFIG_BT.

File: drivers/net/goldfish_rfkill.c [hunk 18]
```diff
@@ -0,0 +1,336 @@
...
+	if (priv->hw_block == hw_block)
+		goto tasklet_done;
+
+	diff = (~priv->hw_block & hw_block) | (priv->hw_block & ~hw_block);
+	priv->hw_block = hw_block;
+
+	ii = ARRAY_SIZE(priv->data);
+	while (ii > 0) {
+		--ii;
+		rfkill = priv->data[ii].rfkill;
```
Comment (tdz) [inline drivers/net/goldfish_rfkill.c]: Just a nit: I'd move this line to line 144, or merge it into line 145.

File: drivers/net/goldfish_rfkill.c [hunk 19]
```diff
@@ -0,0 +1,336 @@
...
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/interrupt.h>
+#include <linux/spinlock.h>
+#include <linux/rfkill.h>
+#include <linux/io.h>
+
+#define PDEV_NAME              "goldfish_rfkill"
+#define RFKILL_NAME_BLUETOOTH  "goldfish-bluetooth"
```
Comment (tdz) [inline drivers/net/goldfish_rfkill.c]: It's only used when CONFIG_BT is set, so it should only be defined then.

File: drivers/net/goldfish_rfkill.c (hunk 13 above)
Comment (vicamo) [inline drivers/net/goldfish_rfkill.c:45]: Landed Android drivers in kernel-3.9 has the same thing.  I'll still keep it.
http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=84d7b768748943db2bb658b43931fdab04c224cc

File: drivers/net/goldfish_rfkill.c (hunk 15 above)
Comment (vicamo) [inline drivers/net/goldfish_rfkill.c]: The logic here doesn't seem to fit a simple if-else branch.  But I can bail-out early instead.

**Comment by vicamo** [comment]
Address above review comments. @tdz do you mind have another review?

**Comment by vicamo** [comment]
Re-enable Bluetooth on x86 after [bug 870422](https://bugzilla.mozilla.org/show_bug.cgi?id=870422).
